### PR TITLE
Allow (internally) to specify custom resolver and glue, and expose some new formula types

### PR DESCRIPTION
**Pull request by ivanradicek**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs [hunk 1]
```diff
@@ -148,7 +151,7 @@ internal static FormulaType Build(DType type)
                     return BindingError;
 
                 default:
-                    throw new NotImplementedException($"Not implemented type: {type}");
+                    throw new NotImplementedException($"Not implemented type: {type} ({type.Kind}) ({type?.GetType().FullName})");
```
Comment (ivanradicek) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:154]: TODO: revert this change. Was useful for debugging.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the MIT license.
+
+using Microsoft.PowerFx.Core.Types;
+
+namespace Microsoft.PowerFx.Types
+{
+    /// <summary>
+    /// Representing UI control type.
+    /// </summary>
+    public class ControlType : AggregateType
```
Comment (ivanradicek) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs:11]: > `: AggregateType`

This will probably not work, as it requires `GetAllNames()` which (I believe) will not work for a control DType.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs [hunk 3]
```diff
@@ -148,7 +151,7 @@ internal static FormulaType Build(DType type)
                     return BindingError;
 
                 default:
-                    throw new NotImplementedException($"Not implemented type: {type}");
+                    return new UnsupportedType(type);
```
Comment (ivanradicek) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:154]: Don't know how sane this is, but it makes it really convenient for what we're trying to do.

Comment (ivanradicek) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:154]: Note that this type does not have a public way of constructing it.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs:11]: >ControlType [](http://example.com/codeflow?start=17&length=11)

Do we need to expose this, or can it fall under Unsupprted?

File: src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs [hunk 5]
```diff
@@ -50,11 +50,21 @@ public IEnumerable<string> GetAllFunctionNames()
         /// </summary>
         /// <param name="alternateConfig">An alternate config that can be provided. Should default to engine's config if null.</param>
         /// <returns></returns>
-        private protected virtual SimpleResolver CreateResolver(PowerFxConfig alternateConfig = null)
+        private protected virtual SimpleResolver CreateSimpleResolver(PowerFxConfig alternateConfig = null)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs:53]: >CreateSimpleResolver [](http://example.com/codeflow?start=49&length=20)

can we avoid the dual chaining virtuals? Maybe just have CreateResolver() return INameResolver? 
I see there are a few places that ask for SimplerResolver, but they can probably just take INameResolver

Comment (ivanradicek) [reply src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs]: Yes, that worked out.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs (hunk 2 above)
Comment (ivanradicek) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/ControlType.cs]: The implementation was dropped.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs (hunk 1 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs]: I'd approve one this change is reverted.

Comment (ivanradicek) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs]: Note that this is an outdated change. All of this is replaced with unsupported type construction.

**Comment by lesaltzm** [comment]
>                 null,

reuse the glue function?

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs:236 in 6b439cc. [](commit_id = 6b439cc33c3f2729b793b81e543a714f935f7cb5, deletion_comment = False)

**Review by lesaltzm — DISMISSED** [review]
:shipit:

**Comment by MikeStall** [comment]
>                 null,

CreateBinderGlue() ? 

---
In reply to: [1129286005](https://github.com/microsoft/Power-Fx/pull/385#issuecomment-1129286005) [](http://example.com/codeflow?ancestors=1129286005,1129286005)

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs:236 in 6b439cc. [](commit_id = 6b439cc33c3f2729b793b81e543a714f935f7cb5, deletion_comment = False)

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/UnsupportedType.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        {
+        }
+
+        /// <inheritdoc />
+        public override void Visit(ITypeVistor visitor)
+        {
+            visitor.Visit(this);
+        }
+
+        internal string Description => $"kind={_type.Kind} fullTypeName={_type.GetType().FullName}";
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/UnsupportedType.cs:24]: >Description [](http://example.com/codeflow?start=24&length=11)

nit: is this just for debugging? If so, you could use [DebuggerDisplay]

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/UnsupportedType.cs (hunk 6 above)
Comment (ivanradicek) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/UnsupportedType.cs]: Done, please do check.

**Review by MikeStall — APPROVED** [review]
